dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/resourcetable/table/CategoryResourceTable.java

```diff
@@ -35,67 +36,82 @@
 import org.hisp.dhis.category.Category;
 import org.hisp.dhis.category.CategoryOptionGroupSet;
 import org.hisp.dhis.commons.util.TextUtils;
+import org.hisp.dhis.db.model.Column;
+import org.hisp.dhis.db.model.DataType;
+import org.hisp.dhis.db.model.Logged;
+import org.hisp.dhis.db.model.Table;
+import org.hisp.dhis.db.model.constraint.Nullable;
 import org.hisp.dhis.resourcetable.ResourceTable;
 import org.hisp.dhis.resourcetable.ResourceTableType;
 import org.hisp.dhis.resourcetable.util.UniqueNameContext;
 
 /**
  * @author Lars Helge Overland
  */
-public class CategoryResourceTable extends ResourceTable<Category> {
+public class CategoryResourceTable implements ResourceTable {
+  private static final String TABLE_NAME = "_categorystructure";
+
+  private final List<Category> categories;
+
   private final List<CategoryOptionGroupSet> groupSets;
 
-  private final String tableType;
+  private final Logged logged;
 
   public CategoryResourceTable(
-      List<Category> objects, List<CategoryOptionGroupSet> groupSets, String tableType) {
-    super(objects);
+      List<Category> categories, List<CategoryOptionGroupSet> groupSets, Logged logged) {
+    this.categories = categories;
     this.groupSets = groupSets;
-    this.tableType = tableType;
+    this.logged = logged;
   }
 
   @Override
-  public ResourceTableType getTableType() {
-    return ResourceTableType.CATEGORY_STRUCTURE;
+  public Table getTable() {
+    return new Table(toStaging(TABLE_NAME), getColumns(), getPrimaryKey(), List.of(), logged);
   }
 
-  @Override
-  public String getCreateTempTableStatement() {
-    String statement =
-        "create "
-            + tableType
-            + " table "
-            + getTempTableName()
-            + " ("
-            + "categoryoptioncomboid bigint not null, "
-            + "categoryoptioncomboname varchar(255), ";
+  private List<Column> getColumns() {
+    List<Column> columns =
+        Lists.newArrayList(
+            new Column("categoryoptioncomboid", DataType.BIGINT, Nullable.NOT_NULL),
+            new Column("categoryoptioncomboname", DataType.VARCHAR_255));
 
     UniqueNameContext nameContext = new UniqueNameContext();
```

Comment:
I wonder if something like the unique name task can and should become part of the builder?!

Reply:
This part is very specific to the resource table, and we depend on the existing table names in analytics, so fits better here.